snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/EclExpression.java

```diff
@@ -122,26 +123,29 @@ public Expression apply(Expression it) {
 				});
 	}
 	
-	public Promise<Set<String>> resolveToConceptsWithGroups(final BranchContext context) {
+	public Promise<Multimap<String, Integer>> resolveToConceptsWithGroups(final BranchContext context) {
 		if (conceptsWithGroups == null) {
 			conceptsWithGroups = SnomedRequests.prepareSearchRelationship()
 					.all()
 					.filterByActive(true)
 					.filterByCharacteristicTypes(SnomedEclRefinementEvaluator.ALLOWED_CHARACTERISTIC_TYPES)
 					.filterBySource(ecl)
 					.filterByGroup(1, Integer.MAX_VALUE)
-					.setFields(SnomedRelationshipIndexEntry.Fields.ID, SnomedRelationshipIndexEntry.Fields.SOURCE_ID)
+					.setFields(SnomedRelationshipIndexEntry.Fields.ID, SnomedRelationshipIndexEntry.Fields.SOURCE_ID, SnomedRelationshipIndexEntry.Fields.GROUP)
 					.build(context.id(), context.branchPath())
 					.execute(context.service(IEventBus.class))
-					.then(new Function<SnomedRelationships, Set<String>>() {
+					.then(new Function<SnomedRelationships, Multimap<String, Integer>>() {
 						@Override
-						public Set<String> apply(SnomedRelationships input) {
-							return FluentIterable.from(input).transform(SnomedRelationship::getSourceId).toSet();
+						public Multimap<String, Integer> apply(SnomedRelationships input) {
+							final Multimap<String, SnomedRelationship> relationshipsBySource = Multimaps.index(input, SnomedRelationship::getSourceId);
+							return Multimaps.transformValues(relationshipsBySource, SnomedRelationship::getGroup);
```

Comment:
`transformValues` returns a `Multimap` view that allows duplication, see the example below:

```
Multimap<String, String> m = ImmutableSetMultimap.of(
    "a", "123",
    "b", "1234",
    "a", "abc", 
    "a", "d");

Multimaps.transformValues(m, String::length) --> {a=[3, 3, 2], b=[4]}
```

To avoid this, wrap the result of `transformValues` in an `ImmutableSetMultimap` before returning it.